cdap-app-templates/cdap-program-report/cdap-program-report1_2.10/src/main/scala/co/cask/cdap/report/ReportAggregationFunction.scala

```diff
@@ -0,0 +1,256 @@
+    val running = statusTimeMap.get("RUNNING")
+    // Get the earliest status with one of the ending statuses
+    val end = statusTimeMap.filterKeys(END_STATUSES.contains).values
+      .reduceOption(Math.min(_, _)) // avoid compilation error with Math.min(_, _) instead of Math.min
+    val startInfo = Option(bufferRow.getAs[Row](Constants.START_INFO))
+    val duration = end.flatMap(e => start.map(e - _))
+    val runtimeArgs = startInfo.map(_.getAs[Map[String, String]](Constants.RUNTIME_ARGUMENTS))
+    val startMethod = getStartMethod(runtimeArgs).name()
+    val r = Row(bufferRow.getAs[String](Constants.NAMESPACE),
+      startInfo.map(_.getAs[String](Constants.ARTIFACT_NAME)).orNull,
+      startInfo.map(_.getAs[String](Constants.ARTIFACT_VERSION)).orNull,
+      startInfo.map(_.getAs[String](Constants.ARTIFACT_SCOPE)).orNull,
+      bufferRow.getAs[String](Constants.APPLICATION_NAME),
+      bufferRow.getAs[String](Constants.APPLICATION_VERSION),
+      bufferRow.getAs[String](Constants.PROGRAM_TYPE), bufferRow.getAs[String](Constants.PROGRAM),
+      bufferRow.getAs[String](Constants.RUN), status,
+      start, running, end, duration, startInfo.map(_.getAs[String](Constants.USER)).orNull,
+      startMethod, runtimeArgs.orNull, 0, 0, 0)
+    LOG.trace("RecordBuilder = {}", buffer)
+    LOG.trace("Record = {}", r)
+    r
+  }
+
+  /**
+    * Returns how the program run was started
+    *
+    * @param runtimeArgs the runtime arguments of the program run
+    * @return one of the methods [[ProgramRunStartMethod.MANUAL]], [[ProgramRunStartMethod.SCHEDULED]]
+    *         and [[ProgramRunStartMethod.TRIGGERED]] each indicating that the program run
+    *         was started manually, scheduled by time, or triggered by certain condition such as new dataset partition
+    *         and program status.
+    */
+  private def getStartMethod(runtimeArgs: Option[scala.collection.Map[String, String]]): ProgramRunStartMethod = {
+    if (runtimeArgs.isEmpty) return ProgramRunStartMethod.MANUAL
+    val scheduleInfoJson = runtimeArgs.get.get(SCHEDULE_INFO_KEY)
+    if (scheduleInfoJson.isEmpty) return ProgramRunStartMethod.MANUAL
+    val scheduleInfo: TriggeringScheduleInfo = GSON.fromJson(scheduleInfoJson.get, classOf[TriggeringScheduleInfo])
+    val triggers = scheduleInfo.getTriggerInfos
+    if (Option(triggers).isEmpty || triggers.isEmpty) return ProgramRunStartMethod.MANUAL
+    triggers.get(0).getType match {
+      case TriggerInfo.Type.TIME => ProgramRunStartMethod.SCHEDULED
+      case _ => ProgramRunStartMethod.TRIGGERED
+    }
+  }
+
+  /**
+    * Converts the row with schema [[INPUT_START_INFO_SCHEMA]] from an input row to a row
+    * with with schema [[BUFFER_START_INFO_SCHEMA]] in the buffer.
+    *
+    * @param startInfoRow the original row with schema [[INPUT_START_INFO_SCHEMA]]
+    * @return a row with schema [[BUFFER_START_INFO_SCHEMA]]
+    */
+  private def convertInputStartInfoRow(startInfoRow: Row): Row = {
```

Comment:
I'll probably remove this method and make the buffer use the same schema as the input row. Attempted this change but the test hasn't passed